gapic/src/main/com/google/gapid/util/Paths.java

```diff
@@ -571,6 +565,64 @@ public static String toString(Object path) {
     return toAny(head);
   }
 
+  public static class CommandFilter {
+    public boolean showHostCommands;
+    public boolean showSubmitInfoNodes;
+    public boolean showSyncCommands;
+    public boolean showBeginEndCommands;
+
+    public CommandFilter(boolean showHostCommands, boolean showSubmitInfoNodes,
+        boolean showSyncCommands, boolean showBeginEndCommands) {
+      this.showHostCommands = showHostCommands;
+      this.showSubmitInfoNodes = showSubmitInfoNodes;
+      this.showSyncCommands = showSyncCommands;
+      this.showBeginEndCommands = showBeginEndCommands;
+    }
+
+    public static CommandFilter fromSettings(Settings settings) {
+      SettingsProto.UI.CommandFilter filter = settings.ui().getCommandFilter();
+      return new CommandFilter(filter.getShowHostCommands(), filter.getShowSubmitInfoNodes(),
+          filter.getShowSyncCommands(), filter.getShowBeginEndCommands());
+    }
+
+    public CommandFilter copy() {
+      return new CommandFilter(
+          showHostCommands, showSubmitInfoNodes, showSyncCommands, showBeginEndCommands);
+    }
+
+    public boolean update(CommandFilter from) {
+      boolean changed =
+          showHostCommands != from.showHostCommands ||
+          showSubmitInfoNodes != from.showSubmitInfoNodes ||
+          showSyncCommands != from.showSyncCommands ||
+          showBeginEndCommands != from.showBeginEndCommands;
+      if (changed) {
+        showHostCommands = from.showHostCommands;
+        showSubmitInfoNodes = from.showSubmitInfoNodes;
+        showSyncCommands = from.showSyncCommands;
+        showBeginEndCommands = from.showBeginEndCommands;
+      }
+      return changed;
```

Comment:
Is this changed value used somewhere else? Or would it matter if we return it true every time we get a command filter update and just assign the values?

Calculating changed is pretty ugly and it will only get uglier with more filters added. 

I can see that this has been called from `SetFilter()` and my assumption is that `setFilter()` is called when filter is actually needed to be changed for example changing a filter in UI. (Please let me know if that's not the case) So based on that assumption(if it is true), every time we update the filter returning `true` might be much cleaner.

Reply:
Yes it is used, if the filter doesn't change, the UI is not updated. There is no point in reloading the tree if the filtering options didn't change (i.e. the user opens the filter pop-up, but doesn't end up changing things, or changes them and then changes them back before closing it).